compiler/luci/pass/src/FuseGRUPass.cpp

```diff
@@ -0,0 +1,674 @@
+          return false;
+      }
+    }
+
+    // Find input and hidden FC weights and biases
+    for (auto node : body_nodes)
+    {
+      auto *fc_node = dynamic_cast<luci::CircleFullyConnected *>(node);
+      if (fc_node == nullptr)
+        continue;
+
+      const auto input_node = dynamic_cast<luci::CircleNode *>(fc_node->input());
+      if (input_node == nullptr)
+        return false;
+
+      // For input hidden FullyConnected - input node is CircleInput node
+      if (dynamic_cast<luci::CircleInput *>(input_node) != nullptr)
+      {
+        _weight_ih = dynamic_cast<luci::CircleConst *>(fc_node->weights());
+        _bias_ih = dynamic_cast<luci::CircleConst *>(fc_node->bias());
+      }
+      // For hidden hidden FullyConnected - input node is CircleGather node
+      else if (dynamic_cast<luci::CircleGather *>(input_node) != nullptr)
+      {
+        _weight_hh = dynamic_cast<luci::CircleConst *>(fc_node->weights());
+        _bias_hh = dynamic_cast<luci::CircleConst *>(fc_node->bias());
+      }
+      else
+      {
+        return false;
+      }
+    }
+
+    if (_weight_ih == nullptr or _weight_hh == nullptr)
+      return false;
+  }
+
+  return true;
+}
+
+class FuseGRU final
+{
+public:
+  FuseGRU(const GRUPatternBase *p) : _p(p) {}
+
+public:
+  void apply(void);
+
+private:
+  luci::CircleGRU *create_circle_gru(loco::Graph *graph);
+
+private:
+  const GRUPatternBase *_p;
+};
+
+template <loco::DataType T>
+void copy_values(const luci::CircleConst *node, luci::CircleConst *cloned)
+{
+  assert(T == node->dtype());
+  assert(T == cloned->dtype());
+
+  const auto size = node->size<T>();
+  cloned->size<T>(size);
+  for (uint32_t i = 0; i < size; i++)
+    cloned->at<T>(i) = node->at<T>(i);
+}
+
+luci::CircleConst *clone_circleconst(luci::CircleConst *node, loco::Graph *graph)
+{
+  auto cloned = graph->nodes()->create<luci::CircleConst>();
+
+  if (cloned != nullptr)
+  {
+    // dtype/shape
+    cloned->dtype(node->dtype());
+    cloned->rank(node->rank());
+
+    // values
+    switch (node->dtype())
+    {
+      case loco::DataType::FLOAT32:
+        copy_values<loco::DataType::FLOAT32>(node, cloned);
+        break;
+
+      case loco::DataType::U8:
+        copy_values<loco::DataType::U8>(node, cloned);
+        break;
+
+      case loco::DataType::S8:
+        copy_values<loco::DataType::S8>(node, cloned);
+        break;
+
+      case loco::DataType::S16:
+        copy_values<loco::DataType::S16>(node, cloned);
+        break;
+
+      case loco::DataType::S32:
+        copy_values<loco::DataType::S32>(node, cloned);
+        break;
+
+      case loco::DataType::S64:
+        copy_values<loco::DataType::S64>(node, cloned);
+        break;
+
+      case loco::DataType::BOOL:
+        copy_values<loco::DataType::BOOL>(node, cloned);
+        break;
+
+      default:
+        assert(false);
+    }
+  }
+
+  return cloned;
+}
+
+luci::CircleGRU *FuseGRU::create_circle_gru(loco::Graph *graph)
+{
+  assert(graph);
```

Comment:
If this `graph` and `_p->_pattern_last_node->graph()` are different, please add a not about this.
As I understand, we're looking with multiple sub `graph` objects so `graph` ptr can be different.